qcog_python_client/qcog/pytorch/handler.py

```diff
@@ -0,0 +1,116 @@
+
+Commands get dispatched through the chain of handlers. Each handler
+checks if it can handle the command. If it can, it executes the command.
+If the handler returns a new command, it gets dispatched again though the chain.
+
+Each handler implements an automatic retry mechanism.
+The number of attempts and the time to wait can be configured for each handler.
+
+In case of error, the `revert` method is called to undo any changes made by the handler.
+"""
+
+from __future__ import annotations
+
+import asyncio
+import enum
+from abc import ABC, abstractmethod
+from typing import Generic, Protocol, TypeVar
+
+
+class BoundedCommand(Protocol):
+    """Command type."""
+
+    command: Command
+
+
+CommandPayloadType = TypeVar("CommandPayloadType", bound=BoundedCommand)
+
+
+class Command(enum.Enum):
+    """Command enum."""
+
+    discover = "discover"
+    validate = "validate"
+    upload = "upload"
+    revert_all = "revert_all"
+
+
+class Handler(ABC, Generic[CommandPayloadType]):
+    """Interface for the chain of responsibility pattern."""
+
+    head: Handler  # Reference to the head of the chain
+    next: Handler | None = None
+    executed: bool = False
+    attempts: int = 3
+    retry_after: int = 3
+    command: Command
+
+    @abstractmethod
+    async def handle(self, payload: CommandPayloadType) -> CommandPayloadType | None:
+        """Handle the data."""
+        ...
+
+    @abstractmethod
+    async def revert(self) -> None:
+        """Revert the changes."""
+        ...
+
+    def set_next(self, next_component: Handler, head: Handler) -> Handler:
+        """Set the next component."""
+        self.next = next_component
+        self.head = head
+        return next_component
+
+    async def dispatch(self, payload: CommandPayloadType) -> Handler:
+        """Dispatch the payload through the chain."""
+        # If the command is a `revert_all` command
+        # and the handler has already been executed
+        # we want to revert the state of the handler
+        # and pass the command to the next handler
+        if payload.command == Command.revert_all and self.executed:
+            await self.revert()
+            if self.next:
+                return await self.next.dispatch(payload)
+            return self
+
+        # If the handler matches the command in the payload
+        # Attempt the execution of the command
+        if self.command == payload.command:
+            for i in range(self.attempts):
+                try:
+                    return await execute_and_dispatch_next(self, payload)
+                except Exception as e:
+                    # Try again based on the number of attempts.
+                    # 1 - revert the state of the handler
+                    # 2 - wait for the specified time
+                    # 3 - try again
+                    print(f"Attempt {i}, error: {e}")
+                    await self.revert()
+                    await asyncio.sleep(self.retry_after)
```

Comment:
this should have a backoff so that as you get to later attempts it sleeps longer. The easy way is `(i + 1) * self.retry_after` but there are other ways youc an do it